File: sentimental-credit/credit.py
from cs50 import get_int

def main():
    number = get_int("Number: ")
    if number == 0:
        print("INVALID")
        return
    result = calculate(number)
    length = len(str(number))
    if result % 10 != 0:
        print("INVALID")
    else:
        while number >= 100:
            number = number // 10
        if length == 15 and (number == 34 or number == 37):
            print("AMEX")
        elif (length == 13 or length == 16) and number // 10 == 4:
            print("VISA")
        elif length == 16 and (number >= 51 and number <= 55):
            print("MASTERCARD")
        else:
            print("INVALID")

def calculate(number):
    multiply = False
    sum_mult = 0
    sum = 0
    while number != 0:
        if multiply:
            double = (number % 10) * 2
            sum_mult += (double % 10) + (double // 10)
        else:
            sum += number % 10
        multiply = not multiply
        number = number // 10
    total = sum + sum_mult
    return total

main()

File: credit/credit.c
#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

int main(void)
// Check the validity of a card
{
    long card = get_long("Number: ");
    long card_number = card;
    // a boolean to indicate whether we multiply by 2 or not
    bool multiply = false;
    int sum_mult = 0;
    int sum = 0;
    int length = 0;
    if (card_number == 0)
    {
        printf("INVALID\n");
        return 0;
    }
    while (card_number != 0)
    {
        if (!multiply)
        {
            int num = card_number % 10;
            sum += num;
            multiply = true;
        }
        else
        {
            int product = 2 * (card_number % 10);
            sum_mult += product / 10 + (product % 10);
            multiply = false;
        }
        card_number = card_number / 10;
        length++;
    }
    int total = sum + sum_mult;
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    // Extract starting digits
    long starting_digits = card;
    while (starting_digits >= 100)
    {
        starting_digits /= 10;
    }
    if (length == 15 && (starting_digits == 34 || starting_digits == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 13 && starting_digits / 10 == 4) ||
             (length == 16 && starting_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else if (length == 16 && (starting_digits >= 51 && starting_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}
